File: fp_core/src/main/src/Elements/Animation.cpp
/*
 * Animation.cpp
 *
 *  Created on: May 14, 2012
 *      Author: leo
 */

#include <fruitpunch/Common/Events.h>
#include <fruitpunch/Elements/Animation.h>
#include <stdio.h>

using namespace std;

namespace fp_core {

Animation::Animation() {
	setStartFrame(0);
	setLoop(false);

}

int Animation::calculateDuration() {
	int highestEndFrame = 0;

	// find the tween that has the highest end frame
	vector<boost::shared_ptr<Playable> >::const_iterator it = mChildren.begin();
	for (; it != mChildren.end(); it++) {
		if ((*it)->getEndFrame() > highestEndFrame) {
			highestEndFrame = (*it)->getEndFrame();
		}
	}

	return highestEndFrame + 1;
}

void Animation::renderSingle(int frame) {
	vector<boost::shared_ptr<Playable> >::iterator it = mChildren.begin();
	for (; it != mChildren.end(); it++) {
		(*it)->setCurrentFrame(frame);
	}

}

void Animation::start() {
	Playable::start();

	vector<boost::shared_ptr<Playable> >::iterator it = mChildren.begin();
	for (; it != mChildren.end(); it++) {
		(*it)->start();
	}
}

void Animation::stop() {
	Playable::stop();

	vector<boost::shared_ptr<Playable> >::iterator it = mChildren.begin();
	for (; it != mChildren.end(); it++) {
		(*it)->stop();
	}
}

void Animation::add(boost::shared_ptr<Playable> playable) {
	mChildren.push_back(playable);
	addOrphan(playable);
}

void Animation::onFirstFrame() {
	if (!isPlaying()) {
		return;
	}

	vector<boost::shared_ptr<Playable> >::iterator it = mChildren.begin();
	for (; it != mChildren.end(); it++) {
		(*it)->start();
	}
}

}/* namespace fp_core */

File: fp_py/src/main/fruitpunch/native/elements/animation.py
from fruitpunch.native.elements.playable import Playable
import libfp_core
from fruitpunch.native.native_class import NativeClass

class Animation(Playable):


    def __init__(self):
        NativeClass.__init__(self)
        
        self.store(libfp_core.Animation())        

    # ---------------------------------------------------------------------
    # Native Methods
    # ---------------------------------------------------------------------

    # _____________________________________________________________________

    def calculate_duration(self):
        return self.run_native("calculate_duration")
    # _____________________________________________________________________

    def render_single(self, frame):
        return self.run_native("render_single", frame)

    # ---------------------------------------------------------------------
    # Properties
    # ---------------------------------------------------------------------


NativeClass.class_map[libfp_core.Animation] = Animation